File: python-algo/leetcode/problem_669.py
"""669. Trim a Binary Search Tree
https://leetcode.com/problems/trim-a-binary-search-tree/
"""
from typing import Optional
from common.tree_node import TreeNode


class Solution:
    def trimBST(self, root: Optional[TreeNode], low: int, high: int) -> Optional[TreeNode]:
        if not root:
            return None
        if low <= root.val <= high:
            root.left = self.trimBST(root.left, low, high)
            root.right = self.trimBST(root.right, low, high)
            return root
        if root.val < low:
            return self.trimBST(root.right, low, high)
        return self.trimBST(root.left, low, high)

File: java-algo/src/main/java/com/leetcode/Problem669.java
package com.leetcode;

import com.common.TreeNode;

/**
 * 669. Trim a Binary Search Tree
 * https://leetcode.com/problems/trim-a-binary-search-tree/
 */
public class Problem669 {
    public TreeNode trimBST(TreeNode root, int low, int high) {
        if (root == null) {
            return null;
        }
        if (low <= root.val && root.val <= high) {
            root.left = trimBST(root.left, low, high);
            root.right = trimBST(root.right, low, high);
        } else if (root.val < low) {
            return trimBST(root.right, low, high);
        } else {
            return trimBST(root.left, low, high);
        }
        return root;
    }
}
